mindone/models/dit.py

```diff
@@ -703,6 +710,263 @@ def DiT_S_8(**kwargs):
+            # (N, D) -> (N*T, D)
+            y_temp = y.repeat_interleave(repeats=self.pos_embed.shape[1], dim=0)
+        else:
+            y_spatial, y_temp = None, None
+
+        # text embedding
+        if text_embed is not None:
+            assert (
+                self.condition == "text"
+            ), f"When input y is not None, expect that the condition type equals to `class`, but got {self.condition}"
+            text_embed = self.text_embedder(text_embed.reshape(bs, -1))  # (N, L*D)
+            # (N, D) -> (N*num_frames, D)
+            text_embed_spatial = text_embed.repeat_interleave(repeats=self.temp_embed.shape[1], dim=0)
+            # (N, D) -> (N*T, D)
+            text_embed_temp = text_embed.repeat_interleave(repeats=self.pos_embed.shape[1], dim=0)
+        else:
+            text_embed_spatial, text_embed_temp = None, None
+
+        # spatial encoder
+        c = self.get_condition_embed(t_spatial, y_spatial, text_embed_spatial)
+        for block in self.blocks:
+            x = block(x, c)  # (N, T, D)
+
+        # add time embed
+        _, num_patches, channels = x.shape
+        x = (
+            x.reshape((bs, num_frames, num_patches, channels))
+            .permute((0, 2, 1, 3))
+            .reshape((bs * num_patches, num_frames, channels))
+        )
+        x = x + self.temp_embed
+
+        # temporal encoder
+        c = self.get_condition_embed(t_temp, y_temp, text_embed_temp)
+        for block in self.temp_blocks:
+            x = block(x, c)  # (B*T, F, D)
+        _, num_frames, channels = x.shape
+        x = (
+            x.reshape((bs, num_patches, num_frames, channels))
+            .permute((0, 2, 1, 3))
+            .reshape((bs * num_frames, num_patches, channels))
+        )
+
+        c = self.get_condition_embed(t_spatial, y_spatial, None)
+        x = self.final_layer(x, c)  # (N, T, patch_size ** 2 * out_channels)
+        x = self.unpatchify(x)  # (N, out_channels, H, W)
+        _, channels, h, w = x.shape
+        x = x.reshape((bs, num_frames, channels, h, w))
+        return x
+
+    def forward_with_cfg(self, x, t, y=None, text_embed=None, cfg_scale=4.0):
+        """
+        Forward pass of DiT, but also batches the unconditional forward pass for classifier-free guidance.
+        """
+        # https://github.com/openai/glide-text2im/blob/main/notebooks/text2im.ipynb
+        half = x[: len(x) // 2]
+        combined = ops.cat([half, half], axis=0)
+        model_out = self.construct(combined, t, y=y, text_embed=text_embed)
+        # For exact reproducibility reasons, we apply classifier-free guidance on only
+        # three channels by default. The standard approach to cfg applies it to all channels.
+        # This can be done by uncommenting the following line and commenting-out the line following that.
+        # eps, rest = model_out[:, :self.in_channels], model_out[:, self.in_channels:]
+        eps, rest = model_out[:, :, :3], model_out[:, :, 3:]
```

Comment:
why only keeps 3 channels in the original impl?

Reply:
There is a PR related to it. [the latent channels should be 4 instead of 3](https://github.com/facebookresearch/DiT/pull/12#top). The DiT author explains that cfg on 4 channels is correct, but for reproducibility reasons, they leaves the wrong code and suggest users to revise it manually. I will correct it soon.